crates/compiler/build/src/program.rs

```diff
@@ -762,120 +836,139 @@ fn build_loaded_file<'a>(
+    linking_strategy: LinkingStrategy,
+    build_host_requested: bool,
+    suppress_build_host_warning: bool,
     wasm_dev_stack_bytes: Option<u32>,
     loaded: roc_load::MonomorphizedModule<'a>,
     compilation_start: Instant,
     out_path: Option<&Path>,
 ) -> Result<BuiltFile<'a>, BuildFileError<'a>> {
+    // get the platform path from the app header
     let platform_main_roc = match &loaded.entry_point {
         EntryPoint::Executable { platform_path, .. } => platform_path.to_path_buf(),
         _ => unreachable!(),
     };
 
-    // For example, if we're loading the platform from a URL, it's automatically prebuilt
-    // even if the --prebuilt-platform CLI flag wasn't set.
-    let is_platform_prebuilt = prebuilt_requested || loaded.uses_prebuilt_platform;
-
-    if is_platform_prebuilt && linking_strategy == LinkingStrategy::Surgical {
-        // Fallback to legacy linking if the preprocessed host file does not exist, but a legacy host does exist.
-        let preprocessed_host_path =
-            platform_main_roc.with_file_name(roc_linker::preprocessed_host_filename(target));
-        let legacy_host_path = legacy_host_file(target, &platform_main_roc);
-        if !preprocessed_host_path.exists() && legacy_host_path.exists() {
-            linking_strategy = LinkingStrategy::Legacy;
-        }
-    }
-
-    // the preprocessed host is stored beside the platform's main.roc
-    let preprocessed_host_path = if linking_strategy == LinkingStrategy::Legacy {
-        if target == Target::Wasm32 {
-            // when compiling a wasm application, we implicitly assume here that the host is in zig
-            // and has a file called "host.zig"
-            platform_main_roc.with_file_name("host.zig")
-        } else {
-            legacy_host_file(target, &platform_main_roc)
-        }
-    } else {
-        platform_main_roc.with_file_name(roc_linker::preprocessed_host_filename(target))
-    };
-
-    let output_exe_path = match out_path {
-        Some(path) => {
-            // true iff the path ends with a directory separator,
-            // e.g. '/' on UNIX, '/' or '\\' on Windows
-            let ends_with_sep = {
-                #[cfg(unix)]
-                {
-                    use std::os::unix::ffi::OsStrExt;
-
-                    path.as_os_str().as_bytes().ends_with(&[b'/'])
-                }
+    let output_exe_path = get_exe_path(
+        out_path,
+        app_module_path.as_path(),
+        target,
+        linking_strategy,
+        link_type,
+    );
 
-                #[cfg(windows)]
-                {
-                    use std::os::windows::ffi::OsStrExt;
+    let dll_stub_symbols =
+        roc_linker::ExposedSymbols::from_exposed_to_host(&loaded.interns, &loaded.exposed_to_host);
 
-                    let last = path.as_os_str().encode_wide().last();
+    let legacy_host_path = target.find_legacy_host(&platform_main_roc);
+    let surgical_artifacts = target.find_surgical_host(&platform_main_roc);
```

Comment:
It's a little strange to try to find both of these files when we know it may not be necessary based on the LinkingStartegy, it can also be confusing to those reading the code.